sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,201 @@
+
+  az_pair * const headers_start = get_headers_start(buffer, max_url_size);
+  uint16_t const max_headers = get_headers_max(buffer, headers_start);
+
+  *p_hrb = (az_http_request_builder){ .buffer = buffer,
+                                      .method_verb = method_verb,
+                                      .url = uri_buf,
+                                      .body = { .begin = NULL, .size = 0 },
+                                      .max_url_size = max_url_size,
+                                      .max_headers = max_headers,
+                                      .retry_headers_start = max_headers,
+                                      .headers_end = 0 };
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_set_query_parameter(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value)
+      || name.size == 0 // name can't be empty
+  ) {
+    return AZ_ERROR_ARG;
+  }
+
+  size_t new_url_size;
+  bool const has_value = value.size > 0;
+  {
+    az_span new_url_span = { .begin = p_hrb->url.begin };
+    size_t const extra_chars_size = sizeof((uint8_t)'?') + (has_value ? sizeof((uint8_t)'=') : 0);
+    size_t const name_and_value_size = name.size + value.size;
+    size_t const appended_size = name_and_value_size + extra_chars_size;
+
+    new_url_size = p_hrb->url.size + appended_size;
+
+    // check for integer ovreflows, and finally whether the result would fit
+    if (name_and_value_size < name.size || name_and_value_size < value.size
+        || appended_size < name_and_value_size || appended_size < extra_chars_size
+        || new_url_size < appended_size || new_url_size < p_hrb->url.size
+        || new_url_size > p_hrb->max_url_size) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    new_url_span.size = new_url_size;
+
+    // check whether name or value regions overlap with destination for some reason (unlikely)
+    if (az_const_span_is_overlap(az_span_to_const_span(new_url_span), name)
+        || az_const_span_is_overlap(az_span_to_const_span(new_url_span), value)) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  // Find whether the URL contains '?'character already.
+  // So that we know if we should append "?text" or "&text".
+  // Scan from the end until the first occurrence of '?'.
+  bool first_parameter = true;
+  for (size_t i = p_hrb->url.size; i > 0;) {
+    --i;
+    if (p_hrb->url.begin[i] == '?') {
+      first_parameter = false;
+      break;
+    }
+  }
+
+  // Appent either '?' or '&'
+  p_hrb->url.begin[p_hrb->url.size] = first_parameter ? '?' : '&';
+  ++(p_hrb->url.size);
+
+  // Append parameter name
+  memcpy(p_hrb->url.begin + p_hrb->url.size, name.begin, name.size);
+  p_hrb->url.size += name.size;
+
+  // Append "=<parameter value>", if there's one.
+  if (has_value) {
```

Comment:
If !has_value, then we'd want to remove the query parameter and the logic would be different. You can't just have "?QPNAME" in the URL without and '=' something.

Reply:
I think RFC3986 doesn't specify that, but I agree, it is unlikely that Azure REST API is using it.

I'll add code to return AZ_ERROR_ARG if value is empty.

I disagree that we should remove if the value is empty. There should be a separate function to remove parameters. This way, user's code works in expected manner: set_parameter() is intent to set. remove_parameter() is intent to remove. Should user want to remove when the value is empty, it is not hard to write "if (!empty) { set } else { remove }".
Plus, this allows us to modify the behavior to allow keys without values in the future, if needed.